emma-language/src/main/scala/org/emmalanguage/api/DataBag.scala

```diff
@@ -313,6 +318,36 @@ trait DataBag[A] extends Serializable {
           }
       })._1
   }
+
+  // -----------------------------------------------------
+  // equals, hashCode and toString
+  // -----------------------------------------------------
+
+  // TODO: Do not fetch here. (Use e.g. an outer join, coded as two comprehensions with `exists`)
+  override def equals(o: Any) = o match {
+    case that: DataBag[A] => {
+      val thisVals = fetch()
+      val thatVals = that.fetch()
+      (thisVals.size == thatVals.size) &&
+        (thisVals.toSet diff thatVals.toSet).isEmpty
+    }
+    case _ => false
+  }
+
+  // TODO: Do this with a fold.
+  override def hashCode(): Int =
```

Comment:
Similar to `equals`, this should be kept abstract here and implemented at each `DataBag` implementation. 

For non-local `DataBag` variants, one can use a flavor of [this snippet](https://gist.github.com/aalexandrov/089f0fd4d3de13d37536f035dfe14c19).